request.go

```diff
@@ -85,17 +86,33 @@ func httpRequest(ctx context.Context, client *Client,
 	// Set the status
 	response.StatusCode = resp.StatusCode
 
+	if resp.Body != nil {
+		// Read the body
+		response.BodyContents, response.Error = ioutil.ReadAll(resp.Body)
+	}
 	// Check status code
-	if http.StatusOK != resp.StatusCode {
+	if http.StatusOK == resp.StatusCode {
+		return
+	}
+	// unexpected status, write an error.
+	if response.BodyContents == nil {
+		// no body so just echo status code.
 		response.Error = fmt.Errorf(
 			"status code: %d does not match %d",
 			resp.StatusCode, http.StatusOK,
 		)
 		return
 	}
-
-	// Read the body
-	response.BodyContents, response.Error = ioutil.ReadAll(resp.Body)
-
+	// have a body so map to an error type and add to the error message.
+	errBody := struct {
+		Error string `json:"error"`
+	}{}
+	if err := json.Unmarshal(response.BodyContents, &errBody); err != nil {
+		response.Error = fmt.Errorf("failed to unmarshal mapi error response: %w", err)
```

Comment:
I would return the marshall error - as it should never really occur - but if it did, its still a different expected error

Reply:
Ah yeah good point, I missed that one 👍